### PR TITLE
add dasd group information to userid

### PR DESCRIPTION
**Pull request by wngzhe**
issue:https://github.ibm.com/zvc/planning/issues/9459
code change: add dasd group information to userid definition,

### CODE REVIEW

**Comment by wngzhe** [comment]
test result:

```
[root@rhel-84-115-120-reg-upgrade-test-compute site-packages]# smcli Image_Query_DM  -T WGNZ0043
USER WGNZ0043 LBYONLY 4G 64G G
INCLUDE ZCCDFLT
COMMAND SET VCONFIG MODE LINUX
COMMAND DEFINE CPU 00 TYPE IFL
COMMAND DEF STOR RESERVED 61440M
* DASD_GROUP ECKD:ICICPL
IPL 0100
LOGONBY DEVCORE1
MACHINE ESA 32
SHARE RELATIVE 100
NICDEF 1000 TYPE QDIO DEVICES 3 MACID 68D1BA LAN SYSTEM VSICIC
MDISK 0100 3390 60001 14564 T60200 MR
*DVHOPT LNK0 LOG1 RCM1 SMS0 NPW1 LNGAMENG PWC20230629 CRCXX
```

File: zvmsdk/smtclient.py [hunk 1]
```diff
@@ -678,6 +678,19 @@ def create_vm(self, userid, cpu, memory, disk_list, profile,
                           " <dev>:<dev>" % pcif)
                 raise exception.SDKInvalidInputFormat(msg=errmsg)
             rd += ' --commandPcif %s' % pcif
+        # add dasd group information to userid as comment
+        disk_pool_info = "DASD_GROUP " + CONF.zvm.disk_pool if CONF.zvm.disk_pool else None
+        if len(disk_list) == 1:
```
Comment (SeanHQF) [inline zvmsdk/smtclient.py:683]: Not sure if we need to handle length of disk_list > 1, do we need to change it to ```if len(disk_list) > 1:```?

Comment (wngzhe) [reply zvmsdk/smtclient.py:683]: reference code process :
line 721 in [smtclient.py]
```
        if len(disk_list) == 1:
            disk = disk_list[0]
            if 'format' in disk and disk['format'].lower() == 'swap':
                disk_pool = disk.get('disk_pool') or CONF.zvm.disk_pool
                disk_pool_info = disk_pool
                if disk_pool is None:
                    # if it's vdisk, then create user direct directly
                    vd = disk.get('vdev') or self.generate_disk_vdev(offset=0)
                    disk['vdev'] = vd
                    sizeUpper = disk['size'].strip().upper()
                    sizeUnit = sizeUpper[-1]
                    if sizeUnit != 'M' and sizeUnit != 'G':
                        errmsg = ("%s must has 'M' or 'G' suffix" % sizeUpper)
                        raise exception.SDKInvalidInputFormat(msg=errmsg)

                    if sizeUnit == 'M':
                        size = int(sizeUpper[:-1])
                        if size > 2048:
                            errmsg = ("%s is great than 2048M" % sizeUpper)
                            raise exception.SDKInvalidInputFormat(msg=errmsg)

                    if sizeUnit == 'G':
                        size = int(sizeUpper[:-1])
                        if size > 2:
                            errmsg = ("%s is great than 2G" % sizeUpper)
                            raise exception.SDKInvalidInputFormat(msg=errmsg)

                    rd += ' --vdisk %s:%s' % (vd, sizeUpper)
                    vdisk = disk
```


and reference test cases:
 @mock.patch.object(smtclient.SMTClient, 'add_mdisks')
    @mock.patch.object(smtclient.SMTClient, '_request')
    @mock.patch.object(database.GuestDbOperator, 'add_guest')
    def test_create_vm_multi_no_diskpool(self, add_guest, request, add_mdisks):
        user_id = 'fakeuser'
        cpu = 2
        memory = 1024
        disk_list = [{'size': '1g',
                      'is_boot_disk': True,
                      'disk_pool': 'ECKD:eckdpool1',
                      'format': 'ext3'},
                     {'size': '1g',
                      'disk_pool': 'ECKD:eckdpool1',
                      'format': 'ext3'},
                     {'size': '1g',
                      'format': 'swap'}]
        profile = 'osdflt'
        max_cpu = 10
        max_mem = '4G'
        base.set_conf('zvm', 'default_admin_userid', 'lbyuser1 lbyuser2')
        base.set_conf('zvm', 'user_root_vdev', '0100')
        base.set_conf('zvm', 'disk_pool', None)
        base.set_conf('zvm', 'user_default_share_unit', 0)
        rd = ('makevm fakeuser directory LBYONLY 1G G --cpus 2 '
              '--profile osdflt --maxCPU 10 --maxMemSize 4G --setReservedMem '
              '--logonby lbyuser1:lbyuser2 --ipl 0100')
        self._smtclient.create_vm(user_id, cpu, memory, disk_list, profile,
                                  max_cpu, max_mem, '', '', '', [], {}, '',
                                  [])
        request.assert_called_with(rd)
        add_mdisks.assert_called_with(user_id, disk_list)
        add_guest.assert_called_with(user_id)

Comment (SeanHQF) [reply zvmsdk/smtclient.py:683]: ok, thx

**Review by SeanHQF — APPROVED** [review]

**Comment by jichenjc** [comment]
>>>COMMAND SET VCONFIG MODE LINUX
COMMAND DEFINE CPU 00 TYPE IFL
COMMAND DEF STOR RESERVED 61440M
* DASD_GROUP ECKD:ICICPL
IPL 0100
LOGONBY DEVCORE1
MACHINE ESA 32
SHARE RELATIVE 100
NICDEF 1000 TYPE QDIO DEVICES 3 MACID 68D1BA LAN SYSTEM VSICIC
MDISK 0100 3390 60001 14564 T60200 MR

is it possible to make the DASD GROUP under MDISK statement ?

**Comment by wngzhe** [comment]
> > > > COMMAND SET VCONFIG MODE LINUX
> > > > COMMAND DEFINE CPU 00 TYPE IFL
> > > > COMMAND DEF STOR RESERVED 61440M
> 
>     * DASD_GROUP ECKD:ICICPL
>       IPL 0100
>       LOGONBY DEVCORE1
>       MACHINE ESA 32
>       SHARE RELATIVE 100
>       NICDEF 1000 TYPE QDIO DEVICES 3 MACID 68D1BA LAN SYSTEM VSICIC
>       MDISK 0100 3390 60001 14564 T60200 MR
> 
> 
> is it possible to make the DASD GROUP under MDISK statement ?

let me try

**Comment by wngzhe** [comment]
> > > > COMMAND SET VCONFIG MODE LINUX
> > > > COMMAND DEFINE CPU 00 TYPE IFL
> > > > COMMAND DEF STOR RESERVED 61440M
> 
>     * DASD_GROUP ECKD:ICICPL
>       IPL 0100
>       LOGONBY DEVCORE1
>       MACHINE ESA 32
>       SHARE RELATIVE 100
>       NICDEF 1000 TYPE QDIO DEVICES 3 MACID 68D1BA LAN SYSTEM VSICIC
>       MDISK 0100 3390 60001 14564 T60200 MR
> 
> 
> is it possible to make the DASD GROUP under MDISK statement ?

Tried but failed. following  userid definition part isn't be added in create_vm , "* DASD_GROUP ECKD:ICICPL" comment is the latest part in call request. 
  
```
IPL 0100
  LOGONBY DEVCORE1
  MACHINE ESA 32
  SHARE RELATIVE 100
  NICDEF 1000 TYPE QDIO DEVICES 3 MACID 68D1BA LAN SYSTEM VSICIC
  MDISK 0100 3390 60001 14564 T60200 MR
```

**Comment by wngzhe** [comment]
following the review comments try add comment to the bottom part of userid definiation, when add midisk
but error shows:
`deploy of virtual machine [rhel86_multipath_ext4_eckd-new-116](https://172.26.3.85/icic/index.html#t=virtualMachines.vm_details&vmId=ee17e354-3428-4662-a30e-854af601eeba) on host 172.26.3.165 failed with exception: Build of instance ee17e354-3428-4662-a30e-854af601eeba was re-scheduled: zVM Cloud Connector request failed: {'overallRC': 500, 'modID': 1, 'rc': 4, 'rs': 6, 'errmsg': 'Unexpected internal error in ZVM SDK, error: SMT internal error. Results: {\'overallRC\': 4, \'rc\': 4, \'rs\': 6, \'errno\': 0, \'strError\': \'\', \'response\': ["(Error) ULTGUT0006E CHANGEVM\'s ADD3390 subfunction encountered an unknown operand: --comment"]}.', 'output': ''}`

 add mdisk use  request is 'changevm WGNZ0047 add3390 ICICPL 0100 10g --mode MR ' The changevm defined in changeVM.py, finall actually call the "Image_Disk_Create"  but 'Image_Disk_Create' don't support comment .
```
 [root@rhel-84-115-120-reg-upgrade-test-compute site-packages]# smcli Image_Disk_Create -h
NAME
  Image_Disk_Create

SYNOPSIS
  smcli Image_Disk_Create [-T] image_name [-v] virtual_address [-m] access_mode

DESCRIPTION
  Use Image_Disk_Create to add a disk that is defined in a virtual image's
  directory entry to that virtual image's active configuration.

  The following options are required:
    -T    The userid or image name of the single image for which the disk
          is being created.
    -v    The virtual device address of the disk to be added
    -m    The access mode requested for the disk
            R: Read-only (R/O) access
            RR: Read-only (R/O) access
            W: Write access
            WR: Write access
            M: Multiple access
            MR: Write or any exclusive access
            MW: Write access is allowed to the disk unconditionally
Note:
  Adding the --addRCheader keyword to any smcli call will change the output.
  The first line will now contain return code information.
  The remaining lines will contain the usual response/error data.

  First line format:
    0 0 rs (details) None           -> no errors found. SMAPI reason code(or 0 if none)
    8 rc rs (details) explanation   -> SMAPI error/async detected. SMAPI return code reason code provided
                                       and brief explanation.
    24 0 0 (details) Input error    -> input syntax error, keyword missing or incorrect.
    25 rc errno (details) message   -> smcli processing error code, errno(or 0 if none) and error message.

  Adding the --timeout keyword to any smcli call will change the SMAPI socket timeout value:
    --timeout    Specify the SMAPI socket timeout value in seconds. Example:
                   --timeout 600
```
 so if we want to add comment beyond the create_vm , smcli change needed,

**Review by bjhuangr — APPROVED** [review]

**Comment by bjhuangr** [comment]
@wngzhe please resolve the conflict.

**Comment by wngzhe** [comment]
close for this pr ls low priority ,will reopen it when necessary.

**Pull request closed without merging**